### PR TITLE
rate_limit_quota: [API] Add time unit `month` and `year`

### PR DESCRIPTION
**Pull request by tyxia**
Even though for us (who introduced the rate_limit_quota feature), `month` and `year` unit are too coarse-grained unit as our traffic will be huge and we will update the Envoy config on more frequent basis, there is business need/case for monthly and yearly subscription from @renuka-fernando side.  
More detailed discussion can be found #23844.

Signed-off-by: tyxia <tyxia@google.com>

### CODE REVIEW

**Comment by renuka-fernando** [comment]
I think we have to update this function too. But Month and Year do not have exact seconds, it varies from month to month, year to year, but I wish we do not want exact seconds here. @tyxia wdyt?

https://github.com/envoyproxy/envoy/blob/5c56e45a041a4446411a8dccb5262b12323ff9be/source/extensions/filters/http/ratelimit/ratelimit_headers.cc#L67-L82

File: api/envoy/type/v3/ratelimit_unit.proto [hunk 1]
```diff
@@ -28,4 +28,10 @@ enum RateLimitUnit {
 
   // The time unit representing a day.
   DAY = 4;
+
+  // The time unit representing a month.
+  MONTH = 5;
```
Comment (adisuissa) [inline api/envoy/type/v3/ratelimit_unit.proto:33]: Should these fields be processed by some code?
I'm assuming the code changes aren't too large, so I suggest adding the code as part of this PR.

**Comment by tyxia** [comment]
> I think we have to update this function too. But Month and Year do not have exact seconds, it varies from month to month, year to year, but I wish we do not want exact seconds here. @tyxia wdyt?
> 
> https://github.com/envoyproxy/envoy/blob/5c56e45a041a4446411a8dccb5262b12323ff9be/source/extensions/filters/http/ratelimit/ratelimit_headers.cc#L67-L82

I think there is misunderstanding here. So do you want to include the month and year to existing rate limiting feature (api/envoy/service/ratelimit/v3/rls.proto)  or the new global rate limiting feature I am working on? The PR here is for new rate limiting feature will not use the code you mentioned above.

File: api/envoy/type/v3/ratelimit_unit.proto (hunk 1 above)
Comment (tyxia) [reply api/envoy/type/v3/ratelimit_unit.proto:33]: Currently there is no impl code for this yet. But let's wait for the clarification from @renuka-fernando above first.

**Comment by renuka-fernando** [comment]
> > I think we have to update this function too. But Month and Year do not have exact seconds, it varies from month to month, year to year, but I wish we do not want exact seconds here. @tyxia wdyt?
> > https://github.com/envoyproxy/envoy/blob/5c56e45a041a4446411a8dccb5262b12323ff9be/source/extensions/filters/http/ratelimit/ratelimit_headers.cc#L67-L82
> 
> I think there is misunderstanding here. So do you want to include the month and year to existing rate limiting feature (api/envoy/service/ratelimit/v3/rls.proto) or the new global rate limiting feature I am working on? Or both?
> 
> The PR here is for new rate limiting feature and will not use the code you mentioned above.

I am planning to use the existing rate limit feature (with the Envoy Rate-Limit service: https://github.com/envoyproxy/ratelimit). I think what you refer to as the new global rate limit feature is RLQS, that @markdroth mentioned in this comment https://github.com/envoyproxy/ratelimit/pull/368#pullrequestreview-1142536075.

It is grateful if it is possible to have these units in the existing RLS feature too.
So both if possible 🙂.

**Comment by tyxia** [comment]
> > > I think we have to update this function too. But Month and Year do not have exact seconds, it varies from month to month, year to year, but I wish we do not want exact seconds here. @tyxia wdyt?
> > > https://github.com/envoyproxy/envoy/blob/5c56e45a041a4446411a8dccb5262b12323ff9be/source/extensions/filters/http/ratelimit/ratelimit_headers.cc#L67-L82
> > 
> > 
> > I think there is misunderstanding here. So do you want to include the month and year to existing rate limiting feature (api/envoy/service/ratelimit/v3/rls.proto) or the new global rate limiting feature I am working on? Or both?
> > The PR here is for new rate limiting feature and will not use the code you mentioned above.
> 
> I am planning to use the existing rate limit feature (with the Envoy Rate-Limit service: https://github.com/envoyproxy/ratelimit). I think what you refer to as the new global rate limit feature is RLQS, that @markdroth mentioned in this comment [envoyproxy/ratelimit#368 (review)](https://github.com/envoyproxy/ratelimit/pull/368#pullrequestreview-1142536075).
> 
> It is grateful if it is possible to have these units in the existing RLS feature too. So both if possible 🙂.

Thanks for confirmation! The changes to existing RLS will be done in a separate PR.

File: api/envoy/type/v3/ratelimit_unit.proto (hunk 1 above)
Comment (tyxia) [reply api/envoy/type/v3/ratelimit_unit.proto:33]: @adisuissa , just heard back from @renuka-fernando that he prefer to have it in both rateLimiting designs.
So maybe we can proceed with this one (new RLQS) which doesn't have the impl yet as its baseline imp PR is pending review ?(also, imp for this particular part might not be prioritized to early phase). WDYT?

Comment (adisuissa) [reply api/envoy/type/v3/ratelimit_unit.proto:33]: SGTM.

**Review by adisuissa — APPROVED** [review]
LGTM.
/lgtm api

**Comment by adisuissa** [comment]
As discussed offline, this should be followed by a PR to update the previously defined [Unit](https://github.com/envoyproxy/envoy/blob/a1f8fa57132e286d8a8c26b87a1f6e8773348175/api/envoy/service/ratelimit/v3/rls.proto#L80) enum.

**Pull request merged**